File: Agent/.NET 6.0/Information.cs
using System.Text.Json.Serialization;
using static Technologai.Agent;

namespace Technologai
{
    public enum InformationState
    {
        DRAFT = 0,
        OPEN = 1,
        CLOSED = 2
    }

    public class Information : IComparable<Information>
    {
        private bool _assessmentQueued = false;

        [JsonIgnore]
        public Agent? Agent { get; set; }

        public string Id { get; }
        public string CreatorId { get; }
        public string? WorkerId { get; set; }
        public string TemplateId { get; }
        public InformationState InformationState { get; internal set; }
        public TemplateState TemplateState { get; private set; }


        public Data? Input { get; }
        public Data? Output { get; private set; }
        //public Data? Instruction { get; private set; }


        // TODO History, Signatures, ReadOnly fields ?        

        [JsonConstructor]
        public Information(string id, string creatorId, string? workerId, string templateId, InformationState informationState,
                            TemplateState templateState, Data? input = null, Data? output = null)
        {
            Id = id;
            CreatorId = creatorId;
            WorkerId = workerId;
            TemplateId = templateId;
            InformationState = informationState;
            TemplateState = templateState;
            Input = input;
            Output = output;
        }

        public Information(Agent agent, string templateId, Data? input = null)
            : this(Technologai.Id.Create(agent.Id), agent.Id, null, templateId, InformationState.OPEN, TemplateState.RESTING, input, null)
        {
            Agent = agent;
            WorkerId = Agent.Catalog[TemplateId].MemberId;
        }

        protected internal async Task<bool> Assess()
        {
            if (Agent == null || _assessmentQueued || TemplateState == TemplateState.PROCESSING || InformationState == InformationState.CLOSED) { return false; }

            // Assessments are debounced. Only one assessment can be queued at a time.
            // FIXME: Not Threadsafe

            if (TemplateState == TemplateState.ASSESSING)
            {
                _assessmentQueued = true;

                while (TemplateState == TemplateState.ASSESSING)
                {
                    await Task.Delay(10);
                }
            }

            if (Agent.Catalog.ContainsKey(TemplateId))
            {
                TemplateState = TemplateState.ASSESSING;

                var result = await Agent.Catalog[TemplateId].Assess(this);

                TemplateState = TemplateState.RESTING;

                _assessmentQueued = false;

                return result;
            }

            _assessmentQueued = false;

            return false;
        }

        protected internal async Task Process()
        {
            if (Agent == null) { return; }

            // Only one process can be in progress at a time. We don't queue up another one.
            // FIXME: Not Threadsafe

            if (TemplateState == TemplateState.RESTING && Agent.Catalog.ContainsKey(TemplateId))
            {
                TemplateState = TemplateState.PROCESSING;

                Output = await Agent.Catalog[TemplateId].Process(this);

                InformationState = InformationState.CLOSED;

                TemplateState = TemplateState.RESTING;

                WorkerId = CreatorId;

                await Agent.PublishAsync(this, null);
            }
        }

        public async Task PublishAsync(OutputCallback? callback, string templateId, Data? input = null)
        {
            if (Agent == null) { return; }

            var information = new Information(Agent, templateId, input);
            Agent.Context.Add(information);
            Agent.Context.Spawn(information.Id, Id);
            await Agent.PublishAsync(information, callback);
        }

        public async Task<Data?> Publish(string templateId, Data? input = null)
        {
            if (Agent == null) { return null; }

            var information = new Information(Agent, templateId, input);
            Agent.Context.Add(information);
            Agent.Context.Spawn(information.Id, this.Id);
            return await Agent.Publish(information);
        }

        public int CompareTo(Information? other)
        {
            return object.ReferenceEquals(other, null) ? 1 : ((Id)Id).CompareTo((Id)other.Id);
        }
    }
}

File: Agent/python_agent/technologai_agent/information.py
from typing import Dict, Optional
# from .agent import Agent
from enum import Enum

class InformationState(Enum):
    DRAFT = 0
    OPEN = 1
    CLOSED = 2

class TemplateState(Enum):
    RESTING = 0
    ASSESSING = 1
    PROCESSING = 2

class Information:
    def __init__(self, agent, template_id: str, input_data = None):
        self.agent = agent
        self.id = None  # This should be generated uniquely
        self.creator_id = agent.identity.member_id  # This should be set appropriately
        self.worker_id = None
        self.template_id = template_id
        self.information_state = InformationState.OPEN  # This should be an enum in a real application
        self.template_state = TemplateState.RESTING  # This should be an enum in a real application
        self.input = input_data if input_data else {}
        self.output = {}
        self.assessment_queued = False

    async def assess(self):
        if self.assessment_queued or self.template_state == TemplateState.PROCESSING or self.information_state == InformationState.CLOSED:
            return

        if self.template_id in self.agent.catalog and self.template_state == TemplateState.RESTING:
            self.template_state = TemplateState.ASSESSING
            self.assessment_queued = True
            result = await self.agent.catalog[self.template_id].assess(self)
            self.template_state = TemplateState.RESTING
            return result
        
        self.assessment_queued = False
        return False

    async def process(self):
        if self.template_state == TemplateState.RESTING:
            self.template_state = TemplateState.PROCESSING
            self.output = await self.agent.catalog[self.template_id].process(self)
            self.template_state = TemplateState.RESTING
            self.information_state = InformationState.CLOSED
            await self.agent.publish_async(information=self)  # Send the closed information back to the agent

    async def publish(self, template_id, input_data=None):
        new_info = Information(self.agent, template_id, input_data)
        self.agent.context.add(new_info)
        self.agent.context.spawn(new_info.id, self.id)
        return await self.agent.publish(new_info)

    def __lt__(self, other):
        return self.id < other.id
